lib/Tools/circt-verilog-lsp-server/VerilogServerImpl/VerilogServer.cpp

```diff
@@ -1030,15 +976,77 @@ static llvm::lsp::Range getRange(const mlir::FileLineColRange &fileLoc) {
       llvm::lsp::Position(fileLoc.getEndLine(), fileLoc.getEndColumn()));
 }
 
+// LSP (0-based line, UTF-16 character) -> byte offset into UTF-8 buffer.
+static inline uint32_t lspPositionToOffset(const slang::SourceManager &sm,
+                                           const slang::BufferID &buf,
+                                           const llvm::lsp::Position &pos) {
+  std::string_view text = sm.getSourceText(buf);
+
+  // 1) Find start-of-line (clamp to EOF if line is past end)
+  size_t lineStart = 0;
+  for (int i = 0; i < pos.line && lineStart < text.size(); ++i) {
+    size_t nl = text.find('\n', lineStart);
+    if (nl == std::string_view::npos) {
+      lineStart = text.size();
+      break;
+    }
+    lineStart = nl + 1;
+  }
```

Comment:
Not blocking but I think this will cause performance regression since it's necessary to walk entire text in the worst case (chisel ~often~ sometimes emit a big file so ...).  LLVM sourceManager caches offsets and do the binary search here: 
https://github.com/llvm/llvm-project/blob/19cd5bd350b730da35629de5095764861f70ecee/llvm/lib/Support/SourceMgr.cpp#L104-L118

Slang source manager actually caches offsets as well,  https://github.com/MikePopoloski/slang/blob/d9ac86a723653474548737f55384b0d99f447d92/source/text/SourceManager.cpp#L661 so perhaps we can do binary search on location to determine line? It's totally fine for me to continue with the current implementation if there a tracking issue :)

Reply:
That's a valid point. I suggest we just cache all line offsets per document and then just add the character offsets in-flight. This should be fine especially if we plan to debounce updating the document state anyway.

I implemented a draft of a version of this.